src/tests/add.rs

```diff
@@ -129,3 +132,103 @@ fn add_key_to_session() {
     assert_eq!(new_payload, new_expected);
     keyring.unlink_key(&key).unwrap();
 }
+
+#[test]
+fn add_encrypted_key_to_session() {
+    let mut keyring = Keyring::attach_or_create(SpecialKeyring::Session).unwrap();
+    let master_key = keyring
+        .add_key::<User, _, _>("foo", "bar".as_bytes())
+        .unwrap();
+    let new_payload = Payload::New {
+        format: Some(Format::default()),
+        keytype: MasterKeyType::User,
+        description: Cow::Borrowed("foo"),
+        keylen: 32,
+    };
+
+    let mut enc_key = keyring
+        .add_key::<Encrypted, _, _>("baz", new_payload)
+        .unwrap();
+
+    // A normal payload update fails
+    assert_eq!(
+        enc_key.update("qux".as_bytes()),
+        Err(errno::Errno(libc::EINVAL))
+    );
+
+    keyring.unlink_key(&enc_key).unwrap();
+    keyring.unlink_key(&master_key).unwrap();
+}
+
+#[test]
+#[should_panic(
+    expected = "called `Result::unwrap()` on an `Err` value: Errno { code: 22, description: Some(\"Invalid argument\") }"
+)]
+fn load_encrypted_key_to_session() {
+    let mut keyring = Keyring::attach_or_create(SpecialKeyring::Session).unwrap();
+    let master_key = keyring
+        .add_key::<User, _, _>("foo", "bar".as_bytes())
+        .unwrap();
+    let new_payload = Payload::New {
+        format: Some(Format::default()),
+        keytype: MasterKeyType::User,
+        description: Cow::Borrowed("foo"),
+        keylen: 32,
+    };
+    let enc_key = keyring
+        .add_key::<Encrypted, _, _>("baz", new_payload)
+        .unwrap();
+    let buf = enc_key.read().unwrap();
+
+    let load_payload = Payload::Load {
+        blob: buf.clone(),
+    };
+
+    keyring.unlink_key(&enc_key).unwrap();
+
+    // This should not panic but currently does due to the use
+    // of ByteBuf when encoding the load payload.
```

Comment:
Is there a reason `ByteBuf`s `LowerHex` is used for the `Load` payload type? AFAIK that is what is causing this test to fail.

Reply:
https://github.com/torvalds/linux/blob/master/Documentation/security/keys/trusted-encrypted.rst

> All user level blobs, are displayed and loaded in hex ascii for convenience, and are integrity verified.

In code, it is [read here](https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L242) and [parsed here](https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L685) by [this function](https://github.com/torvalds/linux/blob/cbafe18c71028d5e0ee1626b4776fea5d5824a78/lib/hexdump.c#L44) which only supports lowercase ASCII hex.

Reply:
Sorry, I didn't do a good job of explaining the issue I hit.

https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L162

> ```
> * load [<format>] <master-key name> <decrypted data length>
> *     <encrypted iv + data>
> ```
The [key format](https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L195), [master key description](https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L201), and the [decrypted data length](https://github.com/torvalds/linux/blob/master/security/keys/encrypted-keys/encrypted.c#L220) are parsed before the hex encoded data.

The current payload encoding for the load payload [encodes everything](https://github.com/mathstuf/rust-keyutils/blob/master/src/keytypes/encrypted.rs#L162) but the key command as lower-case hex. This includes the format, master key, and decrypted data length, which AIUI
should not be encoded as lower-case hex.

Reply:
Oh. Let's fix that in the `Payload` enum representation. I may have mixed up which docs/code I was looking at when making the payload type here.

Tests for the various keytypes was on my list after finishing #34, but the CI issues burned me out for a while on that :/ .

Reply:
I suppose a patch to fix the documentation for the `load` format should also be made:

```
keyctl add encrypted name "load hex_blob" ring
```